### PR TITLE
refactor(reallocate): expect final allocation

### PR DESCRIPTION
**Pull request by Rubilmax**
With this change, the allocator role is responsible for ordering input allocations so as to have sufficient liquidity at each supply step

### CODE REVIEW

**Comment by Rubilmax** [comment]
For reference, the gas diff is:
image

TLDR ; surprisingly, this version costs less. It may be explained by the fact that hardhat tests previously were withdrawing all then supplying some and now are only purely "reallocating" so they do less interactions with Morpho on avg

So this result is biased

**Comment by MerlinEgalite** [comment]
I'm not sure I prefer this version and what is the most intuitive

File: src/MetaMorpho.sol
Comment (Jean-Grimal) [inline src/MetaMorpho.sol:347]:
```suggestion
    function reallocate(MarketAllocation[] calldata allocation) external onlyAllocatorRole {
```
Couldn't it just be named allocation ?

**Review by Jean-Grimal — COMMENTED** [review]
I actually like this implementation

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol:347]: How do name each element of the array then?

Comment (Jean-Grimal) [reply src/MetaMorpho.sol:347]: Oh yes mb

**Review by Jean-Grimal — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Comment by QGarchery** [comment]
I think I lack some context on this PR.

> It is still possible to front-run reallocate and donate to the first withdrawn market up to the expected assets + idle so that the vault lacks liquidity. But it requires even more donations from the attacker

Why would it lack liquidity, isn't it just adding more to the idle liquidity ?

> it solves any front-running issues though

It's contradictory with the description "it is still possible to front-run reallocate" no ?

**Review by MerlinEgalite — APPROVED** [review]

**Comment by QGarchery** [comment]
> Actually you're right it's just adding to the idle liquidity

So it does solve any frontrunning issue, right ? Because the only thing the attacker can do is to add to the idle liquidity (or to any market, but that is always possible)

**Comment by Rubilmax** [comment]
> So it does solve any frontrunning issue, right ? Because the only thing the attacker can do is to add to the idle liquidity (or to any market, but that is always possible)

In the end I agree! Removed the PR's description, waiting for @StErMi 's approval

File: src/MetaMorpho.sol
Comment (MerlinEgalite) [inline src/MetaMorpho.sol:348]: to remove i think

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]

**Review by MerlinEgalite — APPROVED** [review]

File: src/MetaMorpho.sol
Comment (QGarchery) [inline src/MetaMorpho.sol:392]: Curious to know why you switched those 2 lines :eyes:

File: src/MetaMorpho.sol
Comment (QGarchery) [inline src/MetaMorpho.sol:385]: It's minor but this does not compute exactly the same thing as if the check was done after the `supply` call: the variable `supplyAssets` is computed using old values of the total supply and the total supply shares

File: src/MetaMorpho.sol
Comment (QGarchery) [inline src/MetaMorpho.sol:355]: This check is now done for both the supply and the withdraw case, but it is redundant in the supply case because we already check the cap (and a market with cap > 0 cannot have an inconsistent asset)

File: src/MetaMorpho.sol
Comment (QGarchery) [inline src/MetaMorpho.sol:345]: We should update this comment: the allocators can now choose freely a list of supply/withdraw operations to do. I think it should just mention that this is the final allocation

**Review by QGarchery — CHANGES_REQUESTED** [review]
The frontrunning "attacks" that it prevent does not seem very scary because the allocators could always reallocate with a margin and that would mean that the attacker donates a lot. At the same time this implementation is pretty nice so I'm not against doing those changes

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol:392]: This is detail, but I feel like:
- emitting the event right after the CALL is cleaner than after having increased the memory variable
- it allows to free the stack from `suppliedShares`

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol:345]: Updated, is it ok?

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol:385]: Why would they be "old"? Old compared to what? These values are not expected to change in this execution context right?

**Comment by MerlinEgalite** [comment]
Since Morpho Blue has been updated the PR is broken now @Rubilmax (the naming of getters have changed)

File: src/MetaMorpho.sol
Comment (StErMi) [inline src/MetaMorpho.sol:383]: I know that these are details, but I think that it's worth growing these habits. 

The `supplyCap == 0` means two things:
1) explicit: the curator has disabled the market, and you cannot supply to it (or the market has never been enabled)
2) implicit the market that you are trying to supply to have not the same asset as the Metamorpho vault

The first one is easy to see, the second one not. A dev comment would be nice

Comment (MerlinEgalite) [reply src/MetaMorpho.sol:383]: Are you suggesting we should we should revert differently depending on the case?

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol]: done in #313

**Review by MerlinEgalite — APPROVED** [review]

File: src/MetaMorpho.sol
Comment (Rubilmax) [reply src/MetaMorpho.sol:382]: I added a comment as suggested

Comment (StErMi) [reply src/MetaMorpho.sol:382]: > Are you suggesting that we should revert differently depending on the case?

this depends mostly on you and what you want to monitor. the function is authed so I assume that the person should be trusted. 

it would be needed if you want to monitor an authed user that is trying to supply to a not authorized (asset different from the vault) market. but you have to be very stupid to do that knowing that it would anyway revert

**Review by MerlinEgalite — APPROVED** [review]

**Comment by MerlinEgalite** [comment]
@Jean-Grimal @QGarchery can you check this one please?

**Comment by Jean-Grimal** [comment]
> @Jean-Grimal @QGarchery can you check this one please?

LGTM. This implementation feels more natural no me. It really matches the function's name

File: src/MetaMorpho.sol
Comment (StErMi) [inline src/MetaMorpho.sol:661]: Nitpick: the function name should match the underlying function's name called. This also solves the name weirdness that right now could seem to imply that you are going to **supply** (as an action) instead of retrieving the value.

**Comment by StErMi** [comment]
Am I wrong, or before, you were calling the `morpho.accrueInterest` during the reallocate iteration? 
Why have you removed it?

**Comment by MerlinEgalite** [comment]
> Am I wrong, or before, you were calling the `morpho.accrueInterest` during the reallocate iteration? Why have you removed it?

This is a PR on top in fact: https://github.com/morpho-org/metamorpho/pull/313

We separated the changes

**Comment by StErMi** [comment]
> > Am I wrong, or before, you were calling the `morpho.accrueInterest` during the reallocate iteration? Why have you removed it?
> 
> This is a PR on top in fact: #313
> 
> We separated the changes

Please let me know when you have finished all the work and which is the final PR I need to review because it seems that #313 is doing quite many changes

**Review by QGarchery — APPROVED** [review]

**Pull request merged**
